File: leetcode/AIcoding/4.GPT-4o/javascript_solutions/2359-find-closest-node-to-given-two-nodes.js
var closestMeetingNode = function(edges, node1, node2) {
    const n = edges.length;
    const graph = new Map();

    for (let i = 0; i < n; i++) {
        if (!graph.has(i)) graph.set(i, []);
        if (edges[i] !== -1) graph.get(i).push(edges[i]);
    }

    const bfs = (start) => {
        const dist = new Array(n).fill(Infinity);
        const queue = [start];
        dist[start] = 0;
        let index = 0;

        while (index < queue.length) {
            const node = queue[index++];
            for (const neighbor of graph.get(node)) {
                if (dist[neighbor] === Infinity) {
                    dist[neighbor] = dist[node] + 1;
                    queue.push(neighbor);
                }
            }
        }
        return dist;
    };

    const dist1 = bfs(node1);
    const dist2 = bfs(node2);

    let minDist = Infinity;
    let resultNode = -1;

    for (let i = 0; i < n; i++) {
        const maxDist = Math.max(dist1[i], dist2[i]);
        if (maxDist < minDist) {
            minDist = maxDist;
            resultNode = i;
        }
    }

    return resultNode;
};

File: leetcode/AIcoding/3.DeepSeek-v3/go_solutions/2359-find-closest-node-to-given-two-nodes.go
func closestMeetingNode(edges []int, node1 int, node2 int) int {
    n := len(edges)
    dist1 := make([]int, n)
    dist2 := make([]int, n)
    for i := range dist1 {
        dist1[i] = -1
        dist2[i] = -1
    }
    dist1[node1] = 0
    dist2[node2] = 0

    var dfs func(int, []int)
    dfs = func(node int, dist []int) {
        next := edges[node]
        if next != -1 && dist[next] == -1 {
            dist[next] = dist[node] + 1
            dfs(next, dist)
        }
    }
    dfs(node1, dist1)
    dfs(node2, dist2)

    minDist := math.MaxInt32
    result := -1
    for i := 0; i < n; i++ {
        if dist1[i] != -1 && dist2[i] != -1 {
            maxDist := max(dist1[i], dist2[i])
            if maxDist < minDist {
                minDist = maxDist
                result = i
            } else if maxDist == minDist && i < result {
                result = i
            }
        }
    }
    return result
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}